Привет! Меня зовут Валерий Гриднев, я провожу ревью твоей работы. Предлагаю общаться на «ты», если не против.

В работе ты увидишь комментарии, помеченные разным цветом.

<div class="alert alert-success">
Зеленым цветом помечены оброзцовые решения. Которые можно использовать в дальнейшем.
</div>

<div class="alert alert-warning">
Желтым цветом отмечено то, что ты можешь улучшить или доработать. Можешь учесть это при выполнении следующих работ или доработать проект сейчас, но это не обязательно.
</div>

<div class="alert alert-danger">
Красным цветом помечены моменты, которые необходимо доработать, чтобы проект был принят. Уверен, что у тебя все получится. Буду ждать твою работу на повторное ревью. 
</div>

Будет комфортнее работать над проектом вместе, если ты будешь **писать об изменениях** в проекте по моим рекомендациям. Для этого выбери для своих комментариев 

<div class="alert alert-block alert-info">заметный цвет</div>

Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее.
___

# Выбор локации для скважины

Допустим, вы работаете в добывающей компании «ГлавРосГосНефть». Нужно решить, где бурить новую скважину.

Вам предоставлены пробы нефти в трёх регионах: в каждом 10 000 месторождений, где измерили качество нефти и объём её запасов. Постройте модель машинного обучения, которая поможет определить регион, где добыча принесёт наибольшую прибыль. Проанализируйте возможную прибыль и риски техникой *Bootstrap.*

Шаги для выбора локации:

- В избранном регионе ищут месторождения, для каждого определяют значения признаков;
- Строят модель и оценивают объём запасов;
- Выбирают месторождения с самым высокими оценками значений. Количество месторождений зависит от бюджета компании и стоимости разработки одной скважины;
- Прибыль равна суммарной прибыли отобранных месторождений.

## Загрузка и подготовка данных

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as st

from sklearn.model_selection import train_test_split #Деление на выборки (тестовый набор данных)
from sklearn.linear_model import LinearRegression #Линейная регрессия
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler


In [2]:
geo_data_1 = pd.read_csv('/datasets/geo_data_0.csv')
geo_data_2 = pd.read_csv('/datasets/geo_data_1.csv')
geo_data_3 = pd.read_csv('/datasets/geo_data_2.csv')

Просмотрим корректность импорта данных

In [3]:
geo_data_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [4]:
geo_data_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


In [5]:
geo_data_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
id         100000 non-null object
f0         100000 non-null float64
f1         100000 non-null float64
f2         100000 non-null float64
product    100000 non-null float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB


Исключим поле "Уникальный идентификатор"

In [6]:
geo_data_1 = geo_data_1.drop(columns='id')
geo_data_2 = geo_data_2.drop(columns='id')
geo_data_3 = geo_data_3.drop(columns='id')

Просмотрим содержимое импортированных данных

In [7]:
geo_data_1.head(5)

f0        f1        f2     product
0  0.705745 -0.497823  1.221170  105.280062
1  1.334711 -0.340164  4.365080   73.037750
2  1.022732  0.151990  1.419926   85.265647
3 -0.032172  0.139033  2.978566  168.620776
4  1.988431  0.155413  4.751769  154.036647

In [8]:
geo_data_2.head(5)

f0         f1        f2     product
0 -15.001348  -8.276000 -0.005876    3.179103
1  14.272088  -3.475083  0.999183   26.953261
2   6.263187  -5.948386  5.001160  134.766305
3 -13.081196 -11.506057  4.999415  137.945408
4  12.702195  -8.147433  5.004363  134.766305

In [9]:
geo_data_3.head(5)

f0        f1        f2     product
0 -1.146987  0.963328 -0.828965   27.758673
1  0.262778  0.269839 -2.530187   56.069697
2  0.194587  0.289035 -5.586433   62.871910
3  2.236060 -0.553760  0.930038  114.572842
4 -0.515993  1.716266  5.899011  149.600746

<div class="alert alert-warning">
    
**Комментарий ревьюера**
    
Думаю согласишься, что не очень презентабельно выглядит) Про столбец правильно)
</div>

<div class="alert alert-block alert-info">

**Комментарий студента**
    
Привела в более удобный вид с комментариями
</div>

<div class="alert alert-success">
    
**Комментарий ревьюера v_2**
     
Спасибо)Теперь отлично)
</div>

### Вывод

В трёх датасетах удалила столбец ID.  Т.к не какой необходимой информации он не несёт

## Обучение и проверка модели

In [10]:
def geo_data_product(df):
    features = df.drop('product', axis=1)
    target = df['product']
    
    return df, features, target

geo_data_1,features_1,target_1 = geo_data_product(geo_data_1)
geo_data_2,features_2,target_2 = geo_data_product(geo_data_2)
geo_data_3,features_3,target_3 = geo_data_product(geo_data_3)

<div class="alert alert-danger">
    
**Комментарий ревьюера**
    
За функцию дополнительный плюс, молодец)
    
Из минусов, все комментарии должны быть в ячейках **markdown**, а не в ячейках с кодом)
    
И еще, стандартизацию мы должны делать после деления на выборки, т.к. никакие данные кроме тренировочных мы не должны показывать при обучении)
    
Пожалуйста исправь)Если есть какие-то вопросы, буду рад ответить)
</div>

<div class="alert alert-block alert-info">

**Комментарий студента**
    
Выполнила
</div>

### Вывод

Функция geo_data_product обрабатывает сразу 3 датасета.
Столбец product определили как целевой признак

Сформировали выборки которые содержат отдельно признаки и отдельно целевые признаки.

<div class="alert alert-danger">
    
**Комментарий ревьюера v_2**
     
Со столбцами всё верное)Но как мы можем увидеть, ты до сих пор стандартизируешь раньше, чем делишь данные)
    
Сначала мы должны сделать сплит на трейн,тест и только потом обучить стандартизатор на трейни и трансформировать обе части)
    
Если что-то непонятно, задавай вопросы)Нужно чтобы ты точно разобралась с этой темой)
</div>

<div class="alert alert-block alert-info">

**Комментарий студента v_2**
    
Выполнила
</div>

2.1. Разбейте данные на обучающую и валидационную выборки в соотношении 75:25.

In [11]:
def train_test(features, target):
    features_train, features_valid, target_train, target_valid = train_test_split(features, target, test_size=0.25, random_state=12345)
    
    return features_train, target_train, features_valid, target_valid

In [12]:
features_train_1, target_train_1, features_valid_1, target_valid_1 = train_test(features_1, target_1)

features_train_2, target_train_2, features_valid_2, target_valid_2 = train_test(features_2, target_2)

features_train_3, target_train_3, features_valid_3, target_valid_3 = train_test(features_3, target_3)

<div class="alert alert-success">
    
**Комментарий ревьюера v_3**
     
Теперь правильно)
    
Чтобы спрятать предупреждения используй библиотеку **warnings**)
    
В ячейках с объявлением функции лучше ничего больше не писать)
</div>

### Вывод

Функция train_test делит все 3 датасета на обучающую и валидационную выборки в соотношении 75:25.

#2.2. Обучите модель и сделайте предсказания на валидационной выборке.
#2.3. Сохраните предсказания и правильные ответы на валидационной выборке.
#2.4. Напечатайте на экране средний запас предсказанного сырья и RMSE модели.

In [13]:
def LinearRegress(features_train,target_train, features_valid, target_valid):
    numeric = ['f0', 'f1', 'f2']
    scaler = StandardScaler()
    features_train = scaler.fit_transform(features_train[numeric])
    features_valid = scaler.transform(features_valid[numeric])
    
    model = LinearRegression(normalize=False) 
    model.fit(features_train, target_train)
    predictions_valid = model.predict(features_valid)

    RMSE = (mean_squared_error(target_valid, predictions_valid))**0.5
    predictions = pd.Series(predictions_valid)
    product_median = sum(predictions) / len(predictions)
    return RMSE, product_median, predictions_valid

In [14]:
RMSE, product_median, predictions_valid_1 = LinearRegress(features_train_1,target_train_1, features_valid_1, target_valid_1)
print("1 регион, RMSE и средний запас предсказанного сырья:", RMSE, product_median)
RMSE, product_median, predictions_valid_2 = LinearRegress(features_train_2,target_train_2, features_valid_2, target_valid_2)
print("2 регион, RMSE и средний запас предсказанного сырья:", RMSE, product_median)
RMSE, product_median, predictions_valid_3 = LinearRegress(features_train_3,target_train_3, features_valid_3, target_valid_3)
print("3 регион, RMSE и средний запас предсказанного сырья:", RMSE, product_median)

1 регион, RMSE и средний запас предсказанного сырья: 37.5794217150813 92.59256778438008
2 регион, RMSE и средний запас предсказанного сырья: 0.8930992867756158 68.7285468954458
3 регион, RMSE и средний запас предсказанного сырья: 40.02970873393434 94.96504596800506


<div class="alert alert-warning">
    
**Комментарий ревьюера v_3**
     
Почему решила еще использовать **MinMaxScaler**? Снова комментарии рядом с кодом, и вызов функции в ячейке с ее объявлением
</div>

<div class="alert alert-block alert-info">

**Комментарий студента v_3**
    
Выполнила
</div>

### Вывод

Провела обучение с помощью линейной регрессии, результат по регионам следующий:

1. RMSE и средний запас предсказанного сырья: (37.5794217150813, 92.59256778438005)
2. RMSE и средний запас предсказанного сырья: (0.8930992867756158, 68.72854689544582)
3. RMSE и средний запас предсказанного сырья: (40.02970873393434, 94.96504596800509)

Средний запас сырья в равных долях, Среднеквадратическая ошибка в 1 и 3 регионах в большей степени.



<div class="alert alert-success">
    
**Комментарий ревьюера v_3**
     
Да, всё верно)
</div>

## Подготовка к расчёту прибыли

3.1. Все ключевые значения для расчётов сохраните в отдельных переменных.

In [15]:
#Бюджет на разработку скважин в регионе — 10 млрд рублей.
budget = 10000000000
#При нынешних ценах один баррель сырья приносит 450 рублей дохода. 
#Доход с каждой единицы продукта составляет 450 тыс. рублей, поскольку объём указан в тысячах баррелей.
barrel = 450000
#200 лучших точек для разработки
top_wells = 200

3.2. Рассчитайте достаточный объём сырья для безубыточной разработки новой скважины. 

In [16]:
#Сравните полученный объём сырья со средним запасом в каждом регионе.
volume = budget / barrel / top_wells
volume2 = budget / barrel
print("Достаточный объём сырья для безубыточной разработки новой скважины: {:.2f} тыс. баррелей".format(volume))
print("Достаточный объём сырья для безубыточной разработки для региона: {:.2f} тыс. баррелей".format(volume2))
print('Средний запас фактичеcкой нефти в:')
print('1 регионе  : {:.2f}'.format(geo_data_1['product'].mean()))
print('2 регионе  : {:.2f}'.format(geo_data_2['product'].mean()))
print('3 регионе  : {:.2f}'.format(geo_data_3['product'].mean()))

Достаточный объём сырья для безубыточной разработки новой скважины: 111.11 тыс. баррелей
Достаточный объём сырья для безубыточной разработки для региона: 22222.22 тыс. баррелей
Средний запас фактичеcкой нефти в:
1 регионе  : 92.50
2 регионе  : 68.83
3 регионе  : 95.00


При разведке региона исследуют 500 точек, из которых с помощью машинного обучения выбирают 200 лучших для разработки

In [17]:
point = budget / top_wells
print('Средняя прибыль на лучшую точку: ', point, 'тыс.руб.')

Средняя прибыль на лучшую точку:  50000000.0 тыс.руб.


### Вывод

Анализ средних объемов на скважину для каждого региона следующий:

- 1 регионе  : 92.50
- 2 регионе  : 68.83
- 3 регионе  : 95.00

Но для безубыточной разработки новой скважины достаточный объём сырья должен быть 111.11 тыс. баррелей

Достаточный объём сырья для безубыточной разработки для региона: 22222.22 тыс. баррелей

Средняя прибыль на лучшую точку:  50000000 тыс.руб.

<div class="alert alert-warning">
    
**Комментарий ревьюера v_3**
     
Нам нужен запас 111, как в таком случае средние 92 и 95 сали достаточными?)
</div>

<div class="alert alert-block alert-info">

**Комментарий студента v_3**
    
ответ: Вывод был не верным, исправила. Получилось что не одна скважина не имеют достаточных запасов
</div>

<div class="alert alert-success">
    
**Комментарий ревьюера v_4**
     
Поэтому мы и выбираем лучшие скважины из всех) 
</div>

## Расчёт прибыли и рисков 

4.1. Выберите скважины с максимальными значениями предсказаний.

4.2. Просуммируйте целевое значение объёма сырья, соответствующее этим предсказаниям.

4.3. Рассчитайте прибыль для полученного объёма сырья.

In [18]:
def calculation(target, predictions, top_wells, budget):
    predictions = pd.Series(predictions).sort_values(ascending=False)
    selected = target[predictions.index][:top_wells]
    return selected.sum() * barrel - budget

5.1. Примените технику Bootstrap с 1000 выборок, чтобы найти распределение прибыли.

5.2. Найдите среднюю прибыль, 95%-й доверительный интервал и риск убытков. Убыток — это отрицательная прибыль.

5.3. Напишите выводы: предложите регион для разработки скважин и обоснуйте выбор.

In [ ]:
values = []

state = np.random.RandomState(12345)

for i, j in [[target_valid_1, predictions_valid_1], [target_valid_2, predictions_valid_2], [target_valid_3, predictions_valid_3]]:
    for n in range(1000):
        target_subsample = i.reset_index(drop=True).sample(n=500, replace=True, random_state=state)
        predictions_subsample = pd.Series(j)[target_subsample.index]

        values.append(calculation(target_subsample, predictions_subsample, top_wells, budget))

    values = pd.Series(values)
    lower = values.quantile(.025)
    upper = values.quantile(0.975)

    mean = values.mean()
   
    print("Средняя выручка: ", int(mean))
    print("2.5%-квантиль: ", int(lower))
    print("97.5%-квантиль: ", int(upper))
    print("Уровень риска: ", (values < 0).mean() * 100, '%')
    print()
    values = []

<div class="alert alert-success">
    
**Комментарий ревьюера v_3**
     
Наталья, суть работы ты поняла и выполнила практически идеально)Однако нужно исправить оформление, с таким как сейчас, проект я принять не могу) Ответь пожалуйста еще на мои вопросы выше) И конечно задавай свои)Осталось чуть-чуть)
</div>

### Вывод

Провеля анализ мы получили следующие показатели:

- Первый регион имеет среднюю выручку 425938526 и Уровень риска: 6.0 %
- Второй регион имеет среднюю выручку 518259493 и Уровень риска: 0.3 %
- Третий регион имеет среднюю выручку 420194005 и Уровень риска: 6.2 %

Из этого следует, что самая высокая средняя прибыль и самые низкие риски у второго региона. 

Второй регион самый перспективный для разработки.

## Чек-лист готовности проекта

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные подготовлены
- [x]  Выполнен шаг 2: модели обучены и проверены
    - [x]  Данные корректно разбиты на обучающую и валидационную выборки
    - [x]  Модели обучены, предсказания сделаны
    - [x]  Предсказания и правильные ответы на валидационной выборке сохранены
    - [x]  На экране напечатаны результаты
    - [x]  Сделаны выводы
- [x]  Выполнен шаг 3: проведена подготовка к расчёту прибыли
    - [x]  Для всех ключевых значений созданы константы Python
    - [x]  Посчитано минимальное среднее количество продукта в месторождениях региона, достаточное для разработки
    - [x]  По предыдущему пункту сделаны выводы
    - [x]  Написана функция расчёта прибыли
- [x]  Выполнен шаг 4: посчитаны риски и прибыль
    - [x]  Проведена процедура *Bootstrap*
    - [x]  Все параметры бутстрепа соответствуют условию
    - [x]  Найдены все нужные величины
    - [x]  Предложен регион для разработки месторождения
    - [x]  Выбор региона обоснован

<div class="alert alert-success">
    
**Комментарий ревьюера v_4**
     
Наталья, спасибо за работу и исправления. Успехов в дальнейшем обучении)
</div>